yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -534,12 +538,51 @@ private[spark] class Client(
     // send the acl settings into YARN to control who has access via YARN interfaces
     val securityManager = new SecurityManager(sparkConf)
     amContainer.setApplicationACLs(YarnSparkHadoopUtil.getApplicationAclsForYarn(securityManager))
+    val isEncryptedShuffle = if (sparkConf != null) {
+      sparkConf.getBoolean("spark.encrypted.shuffle", false)
+    } else {
+      false
+    }
+    if (isEncryptedShuffle) {
+      setSparkShuffleTokens(credentials, amContainer)
+      // amContainer.setEnvironment()  set the LD_LIBRARY_PATH like "xxx.so" here
+    }
     setupSecurityToken(amContainer)
     UserGroupInformation.getCurrentUser().addCredentials(credentials)
 
     amContainer
   }
 
+  def setSparkShuffleTokens(credentials:Credentials,amContainer:ContainerLaunchContext){
```

Comment:
A lot of the code here seems very similar to code in `ApplicationMaster.scala`. Could you move the similar parts to some helper function somewhere?